src/components/mod.rs

```diff
@@ -0,0 +1,85 @@
+use std::{marker::PhantomData, vec};
+use stwo_prover::core::{
+    channel::Channel,
+    fields::{qm31::SecureField, secure_column::SECURE_EXTENSION_DEGREE},
+    pcs::TreeVec,
+};
+
+pub mod bytes;
+
+/// Const trait that defines the number of columns in the trace table
+pub trait TraceSize {
+    /// Number of columns in the main trace table
+    const MAIN_COLS: usize;
+    /// Number of columns in the interaction trace table
+    const INTERACTION_COLS: usize;
+
+    // Compile-time validation
+    const ASSERT: () = assert!(Self::MAIN_COLS > 0 && Self::INTERACTION_COLS > 0);
+}
+
+/// Generic Claim to implement for claims each component
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct Claim<T: TraceSize> {
+    pub log_size: u32,
+    _marker: PhantomData<T>,
+}
+
+impl<T: TraceSize> Claim<T> {
+    pub const fn new(log_size: u32) -> Self {
+        _ = T::ASSERT;
+
+        Self {
+            log_size,
+            _marker: PhantomData,
+        }
+    }
+
+    #[inline(always)]
+    pub const fn main_cols() -> usize {
+        T::MAIN_COLS
+    }
+
+    #[inline(always)]
+    pub const fn interaction_cols() -> usize {
+        T::INTERACTION_COLS
+    }
+
+    #[inline]
+    pub fn log_sizes(&self) -> TreeVec<Vec<u32>> {
+        TreeVec::new(vec![
+            vec![self.log_size],
+            vec![self.log_size; T::MAIN_COLS],
+            vec![self.log_size; SECURE_EXTENSION_DEGREE * T::INTERACTION_COLS],
+        ])
+    }
```

Comment:
The log_sizes method creates new vectors on each call, which could be inefficient if called frequently.
```suggestion
    #[inline]
    pub fn log_sizes(&self) -> TreeVec<Vec<u32>> {
        if let Some(ref sizes) = self.cached_log_sizes {
            return sizes.clone(); // or maybe just a reference?
        }
        TreeVec::new(vec![
            vec![self.log_size],
            vec![self.log_size; T::MAIN_COLS],
            vec![self.log_size; SECURE_EXTENSION_DEGREE * T::INTERACTION_COLS],
        ])
        self.cached_log_sizes = Some(sizes.clone());
        sizes // or maybe just a reference?
    }
```
Consider caching the result:

Reply:
this is called only once

Reply:
but won't it be called every time a proof is created, or is it expected to be cached by the proving service?